pkg/transport/generic_transport_controller.go

```diff
@@ -0,0 +1,512 @@
+}
+
+// isObjectBeingDeleted is a helper function to check if object is being deleted.
+func isObjectBeingDeleted(object metav1.Object) bool {
+	return !object.GetDeletionTimestamp().IsZero()
+}
+
+func (c *genericTransportController) deleteWrappedObjectsAndFinalizer(ctx context.Context, placementDecision *v1alpha1.PlacementDecision) error {
+	for _, destination := range placementDecision.Spec.Destinations { // TODO need to revisit the destination struct and see how to use it properly
+		if err := c.transportClient.Resource(c.wrappedObjectGVR).Namespace(destination.ClusterId).Delete(ctx, fmt.Sprintf("%s-%s", placementDecision.GetName(), c.wdsName),
+			metav1.DeleteOptions{}); err != nil { // wrapped object name is in the format (PlacementDecision.GetName()-WdsName). see updateWrappedObject func for explanation.
+			if !strings.HasSuffix(err.Error(), notFoundErrorSuffix) { // if object is already not there, we do not report an error cause desired state was achieved.
+				return fmt.Errorf("failed to delete wrapped object '%s' in destination WEC with namespace '%s' - %w", fmt.Sprintf("%s-%s", placementDecision.GetName(),
+					c.wdsName), destination.ClusterId, err)
+			}
+		}
+	}
+
+	if err := c.removeFinalizerFromPlacementDecision(ctx, placementDecision); err != nil {
+		return fmt.Errorf("failed to remove finalizer from PlacementDecision object '%s' - %w", placementDecision.GetName(), err)
+	}
+
+	return nil
+}
+
+func (c *genericTransportController) removeFinalizerFromPlacementDecision(ctx context.Context, placementDecision *v1alpha1.PlacementDecision) error {
+	return c.updatePlacementDecision(ctx, placementDecision, func(placementDecision *v1alpha1.PlacementDecision) bool {
+		return removeFinalizer(placementDecision, transportFinalizer)
+	})
+}
+
+func (c *genericTransportController) addFinalizerToPlacementDecision(ctx context.Context, placementDecision *v1alpha1.PlacementDecision) error {
+	return c.updatePlacementDecision(ctx, placementDecision, func(placementDecision *v1alpha1.PlacementDecision) bool {
+		return addFinalizer(placementDecision, transportFinalizer)
+	})
+}
+
+func (c *genericTransportController) updateWrappedObjectsAndFinalizer(ctx context.Context, placementDecision *v1alpha1.PlacementDecision) error {
+	if err := c.addFinalizerToPlacementDecision(ctx, placementDecision); err != nil {
+		return fmt.Errorf("failed to add finalizer to PlacementDecision object '%s' - %w", placementDecision.GetName(), err)
+	}
+
+	objectsToPropagate, err := c.getObjectsFromWDS(ctx, placementDecision)
+	if err != nil {
+		return fmt.Errorf("failed to get objects to propagate to WECs from PlacementDecision object '%s' - %w", placementDecision.GetName(), err)
+	}
+
+	wrappedObject, err := convertObjectToUnstructured(c.transport.WrapObjects(objectsToPropagate))
+	if err != nil {
+		return fmt.Errorf("failed to wrap objects to a single wrapped object - %w", err)
+	}
+	// wrapped object name is (PlacementDecision.GetName()-WdsName).
```

Comment:
This name composition scheme has two problems.
1. Length. What guarantees that the composite name is not too long? I do not see anything. One solution would be to use object UIDs instead of names.
2. Uniqueness. Both of the input names are under use control and can contain a dash ("-"), if I understand correctly. That means that concatenating them with a dash in between is not guaranteed to create unique names. I think that you could avoid this problem by choosing a separator that is not a valid charater to appear in `c.wdsName` (these are fairly restricted, right? Where is it documented?).

Reply:
under the basic rule that wdsName is unique, the concatenated name suffix is `-unique_suffix`.
unless really unreasonable names are chosen, we get uniqueness.
of course someone can choose unreasonable names but I think that doesn't justify blocking the PR.
we can take care of this in a later PR.

Reply:
Even with unique wdsName values, the concatenations are not guaranteed to be unique. For example: "momma-mia" + "-" + "farrow" == "momma" + "-" + "mia-farrow". I think we should not rely on users to just know that they have to avoid doing this. There is also the length issue.

OK with me to address later, if you have an Issue to do so.

Reply:
I prefer getting this PR merged asap cause it's blocking Maroon in the continuation PR.
minor items like this can be addressed later in very small PR.

Reply:
I have opened Issue #1651 to represent the need to fix this later.